### PR TITLE
Update to support ruby 3

### PR DESCRIPTION
**Pull request by dkubb**
### Summary

This branch upgrades the GitHub Actions build pipeline to test with more modern versions of ruby. It also upgrades some dependencies and adds full simplecov and mutant support, along with artifact exports.

### CODE REVIEW

File: lib/memoizable/memory.rb [hunk 1]
```diff
@@ -96,8 +97,7 @@ def marshal_dump
     #
     # @api public
     def marshal_load(hash)
-      initialize
-      @memory.marshal_load(hash)
+      initialize(hash)
```
Comment (dkubb) [inline lib/memoizable/memory.rb:100]: __comment:__ This does not seem correct.I think I need to check to see what Hash#maeshal_load does, but I suspect it has semantics similar to if this method was:

```ruby
def marshal_load(hash)
  @monitor.synchronize do
    @memory.replace(hash)
  end
  self
end
```

File: lib/memoizable/memory.rb [hunk 2]
```diff
@@ -7,13 +7,13 @@ class Memory
 
     # Initialize the memory storage for memoized methods
     #
-    # @param [ThreadSafe::Cache] memory
+    # @param [Hash] memory
```
Comment (dkubb) [inline lib/memoizable/memory.rb:10]: __TODO:__ Extract a commit to remove this prior to adding it back as a `Hash`, since the original code did not accept an argument.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -11,17 +11,18 @@ jobs:
...
       with:
-        ruby-version: ${{ matrix.ruby }}
-    - run: bundle install
-    - run: bundle exec rspec spec
+        ruby-version: ${{ matrix.ruby-version }}
+        bundler-cache: true # 'ruby/setup-ruby' action provides built-in caching
+    - name: Run tests
+      run: bundle exec rspec
+    - name: Run mutant
+      run: bundle exec mutant run
```
Comment (dkubb) [inline .github/workflows/ci.yml:30]: __TODO:__ Extract mutant gem addition and config to pre-commit.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -11,17 +11,18 @@ jobs:
     - name: Check Whitespace
       run: git diff --check -- HEAD~1
   rspec:
-    name: RSpec
     runs-on: ubuntu-latest
-    timeout-minutes: 5
```
Comment (dkubb) [inline .github/workflows/ci.yml:16]: __TODO:__ Extract these two removals into a pre-commit.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -11,17 +11,18 @@ jobs:
...
+        ruby-version: ['2.7', '3.0', '3.1', '3.2']
     steps:
     - uses: actions/checkout@v2
     - uses: ruby/setup-ruby@v1
       with:
-        ruby-version: ${{ matrix.ruby }}
-    - run: bundle install
-    - run: bundle exec rspec spec
+        ruby-version: ${{ matrix.ruby-version }}
+        bundler-cache: true # 'ruby/setup-ruby' action provides built-in caching
```
Comment (dkubb) [inline .github/workflows/ci.yml:24]: __TODO:__ Extract this configuration change into a pre-commit.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -11,17 +11,18 @@ jobs:
...
       run: git diff --check -- HEAD~1
   rspec:
-    name: RSpec
     runs-on: ubuntu-latest
-    timeout-minutes: 5
     strategy:
       fail-fast: false
       matrix:
-        ruby: ['2.1', '2.3', '2.4', '2.5', '2.6', '2.7', 'jruby']
+        ruby-version: ['2.7', '3.0', '3.1', '3.2']
```
Comment (dkubb) [inline .github/workflows/ci.yml:18]: __TODO:__ Extract this configuration option rename into a pre-commit.

File: .ruby-version [hunk 7]
```diff
@@ -0,0 +1 @@
+3.2.2
```
Comment (dkubb) [inline .ruby-version:1]: __TODO:__ Extract this file addition into a pre-commit.

File: Gemfile [hunk 8]
```diff
@@ -4,9 +4,14 @@ source 'https://rubygems.org'
 
 gemspec
 
-gem 'rake', '~> 10.1'
+group :development, :test do
+  gem 'mutant',       '~> 0.11', '>= 0.11.22'
+  gem 'mutant-rspec', '~> 0.11', '>= 0.11.22'
+  gem 'rake',         '~> 13.0', '>= 13.0.6'
```
Comment (dkubb) [inline Gemfile:10]: __TODO:__ Extract rake upgrade *and* moving it into a group into (2) pre-commits.

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -1,25 +1,22 @@
 # encoding: utf-8
 
-if RUBY_ENGINE.eql?('ruby')
```
Comment (dkubb) [inline spec/spec_helper.rb:3]: __TODO:__ Check if this is still required for jruby. If it is, then restore it and make sure jruby is added back to the matrix.

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -1,25 +1,22 @@
...
 require 'memoizable'
 require 'rspec'
+require 'simplecov'
 
 # Require spec support files and shared behavior
-Dir[File.expand_path('../{support,shared}/**/*.rb', __FILE__)].each do |file|
-  require file.chomp('.rb')
+Pathname.glob(Pathname(__dir__).join('{shared,support}', '**', '*.rb')).each do |file|
+  require file.sub_ext('').to_s
+end
```
Comment (dkubb) [inline spec/spec_helper.rb:10]: __TODO:__ Extract this refactoring into a pre-commit.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -15,11 +13,15 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        ruby: ['2.1', '2.3', '2.4', '2.5', '2.6', '2.7', 'jruby']
+        ruby-version: ['2.7', '3.0', '3.1', '3.2', 'jruby']
```
Comment (dkubb) [inline .github/workflows/ci.yml:16]: __TODO:__ Rename to ruby-version in pre-commit.

File: lib/memoizable/memory.rb [hunk 12]
```diff
@@ -7,11 +7,13 @@ class Memory
...
     #
+    # @param [Hash] memory
+    #
     # @return [undefined]
     #
     # @api private
-    def initialize
-      @memory  = ThreadSafe::Cache.new
+    def initialize(memory)
+      @memory  = memory
```
Comment (dkubb) [inline lib/memoizable/memory.rb:16]: __TODO:__ This refactoring can be split up into smaller chunks.

For example, the injection of the parameter could happen on it's own, and to start it could inject `ThreadSafe::Cache.new`.

File: lib/memoizable/memory.rb [hunk 13]
```diff
@@ -24,7 +26,7 @@ def initialize
     #
     # @api public
     def [](name)
-      @memory.fetch(name) do
+      fetch(name) do
```
Comment (dkubb) [inline lib/memoizable/memory.rb:34]: __TODO:__ This refactoring could be split out.

File: lib/memoizable/memory.rb [hunk 14]
```diff
@@ -38,12 +40,13 @@ def [](name)
...
-      @memory.compute_if_absent(name) do
-        memoized = false
-        value
+      @monitor.synchronize do
+        if key?(name)
+          fail ArgumentError, "The method #{name} is already memoized"
+        else
+          @memory[name] = value
+        end
       end
```
Comment (dkubb) [inline lib/memoizable/memory.rb:58]: __TODO:__ This refactoring could be split out.

File: lib/memoizable/memory.rb [hunk 15]
```diff
@@ -82,20 +85,19 @@ def key?(name)
     #
     # @api public
     def marshal_dump
-      @memory.marshal_dump
+      @memory
```
Comment (dkubb) [inline lib/memoizable/memory.rb:114]: __TODO:__ This refactoring could be split out.

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -1,25 +1,38 @@
...
     - name: Check Whitespace
       run: git diff --check -- HEAD~1
   rspec:
+    needs: base
     runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:
-        ruby: [2.1, 2.3, 2.4, 2.5, 2.6, 2.7, jruby]
+        ruby-version: [2.1, 2.3, 2.4, 2.5, 2.6, 2.7, '3.0', 3.1, 3.2, head, jruby, jruby-head, truffleruby, truffleruby-head]
```
Comment (dkubb) [inline .github/workflows/ci.yml:17]: __TODO:__ This rename could be split out.

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -1,25 +1,38 @@
...
-
 on: push
-
 jobs:
   base:
     name: Base steps
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
```
Comment (dkubb) [inline .github/workflows/ci.yml:8]: __TODO:__ Upgrading to v3 could be split out.

File: spec/spec_helper.rb [hunk 18]
```diff
@@ -12,14 +15,13 @@
...
 end
 
-require 'memoizable'
-require 'rspec'
-
 # Require spec support files and shared behavior
-Dir[File.expand_path('../{support,shared}/**/*.rb', __FILE__)].each do |file|
-  require file.chomp('.rb')
+Pathname.glob(Pathname(__dir__).join('{shared,support}', '**', '*.rb')).each do |file|
+  require file.sub_ext('').to_s
```
Comment (dkubb) [inline spec/spec_helper.rb:24]: __TODO:__ Extract this refactoring.

File: lib/memoizable/method_builder.rb [hunk 19]
```diff
@@ -38,6 +38,9 @@ def initialize(descendant, method)
 
     # The original method before memoization
     #
+    # @example
+    #    method_builder.original_method  # => :foo
+    #
```
Comment (dkubb) [inline lib/memoizable/method_builder.rb:43]: __TODO:__ Extract this into two commits:

- Add Memoizable::Memory YARD docs
- Add Memoizable::MethodBuilder#original_method YARD docs

File: Gemfile [hunk 20]
```diff
@@ -7,6 +7,15 @@ gemspec
 gem 'rake', '~> 10.1'
 
 group :test do
-  gem 'rspec',     '~> 3.1.0'
+  gem 'rspec',     '~> 3.8', '>= 3.8.0'
```
Comment (dkubb) [inline Gemfile:8]: __TODO:__ Extract rspec upgrade into pre-commit.

File: lib/memoizable/memory.rb [hunk 21]
```diff
@@ -24,7 +24,7 @@ def initialize
     #
     # @api public
     def [](name)
-      @memory.fetch(name) do
+      fetch(name) do
```
Comment (dkubb) [inline lib/memoizable/memory.rb:34]: __comment:__ This seems like it could be a valid mutation, but mutant does not consider this an alive mutation.

File: lib/memoizable/memory.rb [hunk 22]
```diff
@@ -75,7 +75,9 @@ def fetch(name)
     #
     # @api public
     def key?(name)
-      @memory.key?(name)
+      @monitor.synchronize do 
+        @memory.key?(name)
+      end
```
Comment (dkubb) [inline lib/memoizable/memory.rb:102]: __comment:__ It's weird that mutant didn't have a mutation that forced a test to be written here.

File: .github/workflows/ci.yml [hunk 23]
```diff
@@ -13,14 +13,7 @@ jobs:
...
       matrix:
-        ruby:
-        - ruby-2.1
-        - ruby-2.3
-        - ruby-2.4
-        - ruby-2.5
-        - ruby-2.6
-        - ruby-2.7
-        - jruby
+        ruby: [2.1, 2.3, 2.4, 2.5, 2.6, 2.7, jruby]
```
Comment (mbj) [inline .github/workflows/ci.yml]: @dkubb I think the commmit message is wrong: Both before this commit and after commit we used arrays. We just now use a different syntax to express the same array.

I'd suggest a commit message of:

> Refactor github actions to inline array syntax.

File: .github/workflows/ci.yml [hunk 24]
```diff
@@ -9,12 +9,11 @@ jobs:
     - name: Check Whitespace
       run: git diff --check -- HEAD~1
   rspec:
-    runs-on: ${{ matrix.os }}
+    runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:
         ruby: [2.1, 2.3, 2.4, 2.5, 2.6, 2.7, jruby]
-        os: [ubuntu-latest]
```
Comment (mbj) [inline .github/workflows/ci.yml:29]: I actually recommend you add osx, but as a warning: OSX is *much* slower on github actions so I recomend to only test a subset of ruby releases on it to preserve personal sanity when iterating.

File: spec/spec_helper.rb [hunk 25]
```diff
@@ -18,8 +18,8 @@
 require 'rspec'
 
 # Require spec support files and shared behavior
-Dir[File.expand_path('../{support,shared}/**/*.rb', __FILE__)].each do |file|
-  require file.chomp('.rb')
+Pathname.glob(Pathname(__dir__).join('{shared,support}', '**', '*.rb')).each do |file|
```
Comment (mbj) [inline spec/spec_helper.rb:23]: I recommend to sort before requirering, I know that in likely these files have no order dependency but its still more deterministic.

File: .github/workflows/ci.yml [hunk 26]
```diff
@@ -21,4 +21,5 @@ jobs:
       with:
         ruby-version: ${{ matrix.ruby-version }}
         bundler-cache: true # 'ruby/setup-ruby' action provides built-in caching
-    - run: bundle exec rspec
+    - name: Run tests
```
Comment (mbj) [inline .github/workflows/ci.yml:26]: I recommend to interpolate the matrix element used into the name to make the github actions UI list the matrix specific in list views.

File: Gemfile [hunk 27]
```diff
@@ -5,6 +5,9 @@ source 'https://rubygems.org'
 gemspec
 
 group :test do
-  gem 'rspec',     '~> 3.1.0'
-  gem 'simplecov', '~> 0.9.1'
+  gem 'rspec', '~> 3.1.0'
+
+  if RUBY_ENGINE.eql?('ruby') && RUBY_VERSION >= '2.7'
```
Comment (mbj) [inline Gemfile:11]: @dkubb I recommend to simply drop all EOL rubies from the CI this helps a lot should you incorporate OSX builds which are much slower.

File: .github/workflows/ci.yml (hunk 24 above)
Comment (dkubb) [reply .github/workflows/ci.yml:29]: @mbj You think this code could/would behave differently on OSX?

File: Gemfile (hunk 27 above)
Comment (dkubb) [reply Gemfile:11]: @mbj the `RUBY_ENGINE` part is still needed to allow this to work with jruby and truffle ruby (I think, but will verify). It should allow the `RUBY_VERSION` to be removed if I do not target EOL.

I think for now I'm going to keep support for older rubies, but I may revisit this in the future in a separate PR. I was mostly concerned with maintaining parity with the previous release and getting things to pass with modern rubies.

File: Gemfile [hunk 28]
```diff
@@ -4,7 +4,7 @@ source 'https://rubygems.org'
 
 gemspec
 
-group :test do
+group :development, :test do
```
Comment (dkubb) [inline Gemfile:7]: __TODO:__ Extract this into it's own commit.

File: .github/workflows/ci.yml (hunk 24 above)
Comment (mbj) [reply .github/workflows/ci.yml:29]: @dkubb yes. As OSX and MRI have slightly different kernel thread primitives which will evenutally affect the GIL behavior I think for this gem the "ruby-platform" semantics may be broken.

Comment (dkubb) [reply .github/workflows/ci.yml:29]: @mbj ok, that sounds good enough for me. I will remove this specific commit and adds OSX to the OS matrix.

Comment (dkubb) [reply .github/workflows/ci.yml:29]: @mbj ok, you're right this is really slow. It slows down my feedback loop too much for regular development so I don't think I'm going to add it to the normal CI step.

However, I do plan on adding CI configuration in the future where I do both of these things:

- Do a full mutant run *without* `--since`
- Run all tests on all ruby versions
- Run all tests on ubuntu-latest and macOS-latest.
- Run every 24 hours

I plan on removing most of the non-3.0 ruby versions from the "regular" run and then have the full 24 hour run all sorts of combinations.

**Review by mbj — APPROVED** [review]
